convert.py: Hot Fix with VocabFactory Integration

teleprint-me opened the pull request:
#### Overview
This pull request introduces significant improvements to the `convert.py` script in the `convert-py` branch of the `llama.cpp` project. Key changes include the introduction of the `VocabFactory` class for streamlined vocabulary handling, refactoring of the `OutputFile` class for better output management, and various user interface improvements.

#### Key Changes

1. **Added BpeVocab and SentencePieceVocab**: Reintroduced the original `BpeVocab` and `SentencePieceVocab` classes to ensure backward compatibility and support for the original llama models. This addition ensures that existing workflows remain functional and uninterrupted.

2. **New HfVocab Supporting AutoTokenizer Vocab**: Replaced the previous `VocabLoader` with the new `HfVocab` class. This class streamlines the integration of `transformers.AutoTokenizer`, ensuring forward compatibility and more efficient handling of Hugging Face tokenizers.

3. **VocabFactory Implementation**: Implemented a new `VocabFactory` class to encapsulate the logic for handling different types of vocabularies. This enhancement significantly improves the flexibility and maintainability of vocabulary loading and processing, accommodating a wider range of model types and formats.

4. **Argument Parser and CLI Improvements**: The command-line interface has been thoroughly reorganized and refined. Improvements include better argument naming, enhanced error handling, and a more intuitive user interface, all of which contribute to improved usability and a smoother user experience.

5. **Code Quality and Readability**: Dedicated efforts were made to enhance code readability, organization, and documentation throughout the `convert.py` script. These improvements align with best practices in Python development, aiming to facilitate easier maintenance and future enhancements.

The original core llama models should now convert as expected with this hot fix.

[Screencast from 2024-01-07 15-30-21.webm](https://github.com/ggerganov/llama.cpp/assets/77757836/da955db6-3f26-43c6-867b-199f3e988352)

#### Migration Path for `transformers.AutoTokenizer` Support
To further streamline the functionality of `convert.py` and align with the modular design approach, I recommend migrating the support for `transformers.AutoTokenizer` from `convert.py` to a dedicated script, `convert-hf-to-gguf.py`. This migration will allow for more focused handling of Hugging Face models and tokenizers, ensuring each script maintains a clear and distinct purpose. This separation of concerns will not only simplify the maintenance but also enhance the scalability of our tooling for different model formats.

#### Request for Review
I kindly request a review of these changes and welcome any feedback or suggestions. The improvements aim to make the `convert.py` script more robust, flexible, and user-friendly, and I believe they will significantly benefit the `llama.cpp` project.

teleprint-me committed:
Update Imports and Add Notes for Future Reference

- Updated import statements in `convert.py`.
- Added import for `AutoTokenizer` from `transformers` module.
- Added conditional import for `gguf` from the local directory.
- Added comments and notes for future reference.

Additional Notes:

- Noted removal of a redundant `TypeAlias` import.
- Noted the removal of a `gguf` debug statement.
- Commented on the presence of `ARCH` and `NDArray` definitions.
- Commented on cleaning up and refactoring data type definitions.

teleprint-me committed:
Merge branch 'master' into convert-py

teleprint-me committed:
Refine Model Hyperparameters and Params Class

- Updated type annotations to use `Optional` for clarity.
- Improved method names and attribute consistency.
- Removed unnecessary variables for better code readability.

Additional Notes:

- Highlighted the use of `Optional` for clearer intent.
- Ensured backward and forward compatibility.

teleprint-me committed:
Restore BpeVocab and SentencePieceVocab classes

- Restored the BpeVocab class for handling BPE tokenization.
- Restored the SentencePieceVocab class for SentencePiece tokenization.

These classes are essential for maintaining the original behavior of the codebase.

teleprint-me committed:
refactor: Standardize vocabulary handling with HfVocab

- Replaced VocabLoader with HfVocab, aligning vocabulary handling across classes.
- Updated initialization of HfVocab with local_files_only=True for AutoTokenizer.
- Introduced optional parameter fname_added_tokens for flexible added token management.
- Streamlined added token handling for clarity and conciseness.
- Maintained special tokens and IDs, enhancing token management.
- Simplified token processing methods for improved readability.
- Added a placeholder for score computation with a default value of -1000.0.
- Optimized newline token check for efficiency.
- Updated __repr__ function for clarity in representation.
- Adjusted type alias Vocab to include BpeVocab, SentencePieceVocab, and HfVocab.
- Removed redundant code related to special token handling, reverse vocabulary mapping, and vocabulary file detection.

This refactoring promotes a standardized and modular approach to vocabulary management, facilitating future integration with a VocabFactory and improving code maintainability and scalability.

teleprint-me committed:
refactor: Enhance readability, functionality, and code quality

- Improved code formatting and readability for better maintainability.
- Refactored LazyUnpickler's CLASSES dictionary for clarity.
- Added print statements and warnings in check_vocab_size for user feedback.
- Removed find_vocab_file_path, as it's superseded by VocabFactory.
- Preparatory changes for upcoming classes: OutputFile and VocabFactory.
- Overall focus on code quality, error handling, and consistency.

These changes reflect a continuous effort to refine the codebase, ensuring it meets best practices and prepares for future enhancements, such as the VocabFactory.

teleprint-me committed:
refactor: Update OutputFile class for enhanced model vocabulary management

- Restructured the constructor for improved readability.
- Updated `add_meta_arch` method for flexible model name determination.
- Introduced `handle_tokenizer_model` for mapping vocab types to supported tokenizer models.
- Streamlined vocabulary extraction with `extract_vocabulary_from_model`.
- Simplified vocabulary metadata addition using `add_meta_vocab`.
- Refactored `add_tensor_info` for clarity and consistency.
- Improved error handling for better user feedback.

These changes signify the development of a versatile and comprehensive `OutputFile` class, enabling efficient management of model conversion output, metadata, vocabulary, and tensor information.

teleprint-me committed:
feat: Introduce VocabFactory for flexible vocabulary management in model conversion

- The VocabFactory class is added to facilitate modular vocabulary handling.
- The constructor initializes a directory path and detects vocabulary-related files.
- The _select_file method provides file paths based on vocabulary type (e.g., BPE, SentencePiece).
- _create_special_vocab generates special vocabularies, accommodating different types.
- The load_vocab method loads vocabularies, handling BPE, SentencePiece, and Hugging Face Fast Tokenizer.
- Error handling and logging enhance debugging and user feedback.
- The modular and flexible design simplifies vocabulary management and supports future extensions.

The VocabFactory class enhances code modularity and maintainability, allowing versatile vocabulary handling in the model conversion process.

teleprint-me committed:
refactor: Improve code organization, argument parsing, and user interface

- Renamed 'default_outfile' to 'default_output_file' for clarity.
- Refactored argument parser setup into 'get_argument_parser' function.
- Introduced descriptive comments for each argument in the parser.
- Added '--vocab-type' argument with choices ["spm", "bpe", "hfft"] for vocabulary processing.
- Improved flag naming consistency: '--outfile' to '--out-file' and '--bigendian' to '--big-endian'.
- Enhanced error handling to prevent overwriting input data in 'default_output_file'.
- Made 'argv' in 'main' an optional parameter for flexibility.
- Introduced dynamic import for 'awq.apply_awq' based on 'args.awq_path' for conditional dependency.

These changes enhance code clarity, organization, and the user interface of the script, aligning it with Python best practices and improving maintainability.

teleprint-me committed:
refactor: Further refine functionality, improve user interaction, and streamline vocabulary handling

- Renamed command-line arguments for clarity and consistency.
- Improved path resolution and import adjustments for robustness.
- Thoughtfully handled 'awq-path' and conditional logic for the weighted model.
- Enhanced model and vocabulary loading with the 'VocabFactory' class for structured and adaptable loading.
- Strengthened error handling and user feedback for a more user-friendly experience.
- Structured output file handling with clear conditions and defaults.
- Streamlined and organized the 'main' function for better logic flow.
- Passed 'sys.argv[1:]' to 'main' for adaptability and testability.

These changes solidify the script's functionality, making it more robust, user-friendly, and adaptable. The use of the 'VocabFactory' class is a notable enhancement in efficient vocabulary handling, reflecting a thoughtful and iterative approach to script development.

teleprint-me committed:
chore: Apply ruff formatting to convert.py

Signed-off-by: teleprint-me <77757836+teleprint-me@users.noreply.github.com>

ggerganov commented on convert.py:
```diff
@@ -1178,32 +1602,111 @@ def do_dump_model(model_plus: ModelPlus) -> None:
+    )
+
+    parser.add_argument(
+        "--out-type",
```
Update `--outtype` instances in the main README.md with the new `--out-type`

ggerganov commented on convert.py:
```diff
@@ -841,30 +1099,42 @@ def bounded_parallel_map(func: Callable[[In], Out], iterable: Iterable[In], conc
 def check_vocab_size(params: Params, vocab: Vocab, pad_vocab: bool = False) -> None:
     if params.n_vocab != vocab.vocab_size:
         if params.n_vocab == vocab.vocab_size:
```
This is not related to the changes in this PR, but the second `if` is never true

ggerganov commented on convert.py:
```diff
@@ -291,129 +362,254 @@ def loadOriginalParamsJson(model: LazyModel, config_path: Path) -> Params:
     @staticmethod
-    def load(model_plus: ModelPlus) -> Params:
-        hf_config_path   = model_plus.paths[0].parent / "config.json"
+    def load(model_plus: ModelPlus) -> "Params":
```
Is this correctly changed from `Params` to `"Params"`?

cebtenzzre replied:
This is incorrect - we are using `from __future__ import annotations` above, so type annotations have no reason to be quoted.

teleprint-me replied:
@cebtenzzre, I understand your point about the `from __future__ import annotations` directive and how it affects the need for string-based annotations. The use of `"Params"` as a forward reference was intended as a precaution against potential circular dependencies, as outlined in PEP 563 and PEP 649.

However, I respect the established practices of this project. If the consensus is that explicit string annotations are not necessary or preferred in the codebase, I'm willing to adapt to that standard. I'll always appreciate your guidance and expertise on any matter.

ggerganov reviewed (approved):
The vanilla LLaMA v1 and v2 models that I have downloaded now work after updating the `n_vocab` in the `params.json` to 32000 (the original file has -1 for some reason and we generate a warning) so this is good

Btw, in this case it would have been better to first submit a minimal PR that fixes #4493 and a follow-up PR with code-style changes, because it is more difficult to review now

teleprint-me commented:
That's why I formatted it last. It would be easier to rollback if it became an issue. 

I intentionally did individual commits, rather than one large commit, as I usually do, so it would be easier to track the changes and rollback any commits if needed.

Will keep in mind for future. Would definitely recommend a formatter though. Ruff can be customized. 

I tested it with llama, mistral, mixtral, and deepseek. Regardless, glad it's working again.

ggerganov commented:
Ok thanks. After fixing the README and the flake CI, we can merge

teleprint-me committed:
Revert to commit 0614c33

teleprint-me committed:
chore: Apply flake8 formatting rules

Signed-off-by: teleprint-me <77757836+teleprint-me@users.noreply.github.com>

teleprint-me commented:
@ggerganov

Given the uncertainties around the `n_vocab` parameter being set to `-1`, I've taken a cautious approach in this PR to ensure immediate stability. This issue will need further investigation post-release to understand its full impact. It's important to note that the conversion process appears to succeed with `vocab_size` set to either `-1` or `32000`, but the implications of this success are currently unclear.

teleprint-me committed:
refactor: Revise `check_vocab_size` for Enhanced Clarity and Correctness

- Resolved an unreachable branch issue by reorganizing the conditional structure.
- Moved the special case check for `params.n_vocab == -1` to the top for immediate assertion.
- Flattened the conditional logic for improved clarity and predictability of the function's behavior.

These changes enhance the readability and functional correctness of the `check_vocab_size` function without altering its intended functionality.

ggerganov committed:
py : fix outfile and outtype

ggerganov committed:
py : suggest hint for missing vocab size

cebtenzzre commented on convert.py:
```diff
@@ -153,65 +183,85 @@ def type_for_tensor(self, name: str, tensor: LazyTensor) -> DataType:
 
     # path to the directory containing the model files
-    path_model: Path | None = None
+    path_model: Optional[Path] = None
```
Why aren't we using [PEP 604](https://peps.python.org/pep-0604/) types anymore?

cebtenzzre commented on convert.py:
```diff
@@ -1145,30 +1355,95 @@ def load_some_model(path: Path) -> ModelPlus:
+                self.files[file] = parent_file_path
+
+    def _select_file(self, vocabtype: Optional[str]) -> Path:
+        if vocabtype in ["spm", "bpe"]:
+            # For SentencePiece and BPE, return specific files as before
+            file_key = "tokenizer.model" if vocabtype == "spm" else "vocab.json"
+            if self.files[file_key]:
+                return self.files[file_key]
+            else:
+                raise FileNotFoundError(f"{vocabtype} {file_key} not found.")
+        elif vocabtype == "hfft":
+            # For Hugging Face Fast Tokenizer, return the directory path instead of a specific file
+            return self.path
+        else:
+            raise ValueError(f"Unsupported vocabulary type {vocabtype}")
```
Several examples of else-after-return here, which is an anti-pattern.

cebtenzzre commented:
How do we agree as a community on a code style for the Python scripts? I don't want this to become a tug-of-war. I have so far tried to keep my changes minimal (and tried not to be opinionated about whitespace), but this PR reverts many of them for no clear reason.

teleprint-me commented:
@cebtenzzre, I appreciate your point about maintaining a consistent code style across the project. I agree that the best way forward is to decide as a community on a linter and formatter for the Python scripts. This approach would help us avoid any style conflicts and ensure uniformity in the codebase.

My use of PEP-0484 guidelines in this PR was based on habit and not intended to override or challenge previous stylistic choices. I apologize if it seemed to revert any changes unintentionally. 

I'm fully open to adapting to the community's agreed-upon style standards. To facilitate this, I suggest we could include configuration files for the chosen linter and formatter in our repository. This would automate the enforcement of our style rules and make it easier for all contributors to adhere to them consistently.

We can work together to establish any standards and integrate them into our workflow. I'm looking forward to your thoughts and suggestions on this.

ggerganov commented:
I guess now is a good time to apply an auto-formatting tool to the python code as I believe there aren't any major developments going on, so conflicts should be minimal. I don't have preference so open to suggestions

@cebtenzzre The main reason to merge this PR is because the `convert.py` script has been broken for a long time and I wanted to get it fixed. I would've preferred a much smaller diff, but fixing the conversion seemed more important

teleprint-me commented:
`ruff` and `black` are both solid choices for auto-formatting. `ruff` has been gaining popularity recently, while `black` is more established and widely used. Since `flake8` is already part of the linter setup, it might be good to align with that, especially if it's integrated into the CI pipeline.

What do you think @cebtenzzre ?
